### PR TITLE
enable directly setting partitions defs on jobs instead of partitioned config

### PR DESCRIPTION
**Pull request by sryza**
This is useful if you are accessing partition_key exclusively via context, because then specifying empty config is just boilerplate.

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/decorators/job.py [hunk 1]
```diff
@@ -144,6 +148,9 @@ def job(
         version_strategy (Optional[VersionStrategy]):
             Defines how each op (and optionally, resource) in the job can be versioned. If
             provided, memoizaton will be enabled for this job.
+        partitions_def (Optional[PartitionsDefinition]): Defines a screet set of partition keys
```
Comment (prha) [inline python_modules/dagster/dagster/core/definitions/decorators/job.py:151]: is this supposed to be `discrete`?  Or does `screet` mean something I'm not aware of...

File: python_modules/dagster/dagster/core/definitions/graph_definition.py [hunk 2]
```diff
@@ -450,12 +451,15 @@ def to_job(
             version_strategy (Optional[VersionStrategy]):
                 Defines how each solid (and optionally, resource) in the job can be versioned. If
                 provided, memoizaton will be enabled for this job.
+            partitions_def (Optional[PartitionsDefinition]): Defines a screet set of partition keys
```
Comment (prha) [inline python_modules/dagster/dagster/core/definitions/graph_definition.py:454]: same here.

**Review by prha — COMMENTED** [review]
Yeah, I think this makes sense to me.

Is the plan to also add equivalents of `static_partitioned_config` and `dynamic_partitioned_config`?

File: python_modules/dagster/dagster/core/definitions/decorators/job.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/core/definitions/decorators/job.py:151]: Oops, definitely a typo.

**Comment by sryza** [comment]
> Is the plan to also add equivalents of static_partitioned_config and dynamic_partitioned_config?

@prha - yes. Just added.  Doing that made me think that it's a little bit weird to have functions that return these objects instead of just asking users to construct them directly.  So I put together a version that replaces `hourly_partitions_def` with `HourlyPartitionsDefinition`, etc.  And added `StaticPartitionsDefinition` and `DynamicPartitionsDefinition` to the top-level API.  Let me know what you think.

**Review by prha — APPROVED** [review]

**Comment by prha** [comment]
I think this looks great!

**Pull request merged**